include/deal.II/matrix_free/tensor_product_kernels.h

```diff
@@ -3518,12 +3613,29 @@ namespace internal
             const Number2 test_grad_xy =
               dim > 1 ? test_grads_value[0] * shapes[i1][0][1] :
                         test_grads_value[0];
-
-            Number2 *values_ptr = values.data() + i + i1 * length;
+            Number2 test_value_y_2;
+            if (n_values > 1)
+              test_value_y_2 = dim > 1 ?
+                                 test_grads_value[3] * shapes[i1][0][1] :
+                                 test_grads_value[3];
+
+            Number2 *values_ptr = values + i + i1 * length;
+            Number2 *values_ptr_2 =
+              n_values > 1 ? values_ptr + Utilities::pow(length, dim) : nullptr;
             for (unsigned int i0 = 0; i0 < length; ++i0)
               {
-                values_ptr[i0] += shape_values_x[i0] * test_value_y;
+                if (add)
+                  values_ptr[i0] += shape_values_x[i0] * test_value_y;
+                else
+                  values_ptr[i0] = shape_values_x[i0] * test_value_y;
                 values_ptr[i0] += shape_derivs_x[i0] * test_grad_xy;
+                if (n_values > 1)
+                  {
+                    if (add)
```

Comment:
```suggestion
                    // Note that 'add' is a template argument, so the compiler will remove this check
                    if (add)
```

Reply:
I have added a comment in all three functions containing these `if(add)` statements.